examples/async/s3_cp.ml

```diff
@@ -0,0 +1,390 @@
+   that has the "s3:GetObject" and "s3:PutObject" permissions enabled
+   for the bucket you are interacting with.
+
+   As this is an example, straightforwardness is prized. One should
+   not use this for a production system, nor assume that it offers a good
+   example of abstraction, interface design or error handling.
+*)
+
+open Core.Std
+open Async.Std
+open Cohttp
+open Cohttp_async
+
+let ksrt = fun (k,_) (k',_) -> String.compare k k'
+
+module Compat = struct
+  (** Things we need to make this happen that, ideally, we'd like other
+     libraries to provide and that are orthogonal to the example here *)
+
+  let encode_string s =
+    (* Percent encode the path as s3 wants it. Uri doesn't
+       encode $, or the other sep characters in a path.
+       If upstream allows that we can nix this function *)
+    let n = String.length s in
+    let buf = Buffer.create (n * 3) in
+    for i = 0 to (n-1) do
+      let c = String.get s i in
+      match c with
+      | 'a' .. 'z'
+      |'A' .. 'Z'
+      | '0' .. '9'
+      | '_' | '-' | '~' | '.' | '/' -> Buffer.add_char buf c
+      | '%' ->
+        (* Sigh. Annoying we're expecting already escaped strings so ignore the escapes *)
+        begin
+          let is_hex = function
+            | 'a' .. 'z' | 'A' .. 'Z' | '0' .. '9' -> true
+            | _ -> false in
+          if (i + 2) < n then
+            if is_hex(String.get s (i+1)) && is_hex(String.get s (i+2)) then
+              Buffer.add_char buf c
+            else
+              Buffer.add_bytes buf "%25"
+        end
+      | _ -> Buffer.add_bytes buf (Printf.sprintf "%%%X" (Char.to_int c))
+    done;
+    Buffer.contents buf
+
+  let hexa = "0123456789abcdef"
+
+  let of_char c =
+    let x = Char.to_int c in
+    hexa.[x lsr 4], hexa.[x land 0xf]
+
+  let cstruct_to_hex_string cs =
+    let open Cstruct in
+    let n = cs.len in
+    let buf = Buffer.create (n * 2) in
+    for i = 0 to n - 1 do
+      let c = Bigarray.Array1.get cs.buffer (cs.off+i) in
+      let (x,y) = of_char c in
+      Buffer.add_char buf x;
+      Buffer.add_char buf y;
+    done;
+    Buffer.contents buf
+
+  let encode_query_string uri =
+    (* URI.encoded_of_query drops the null values. So if we have foo.com?a
+       will return a, but s3 needs a=. This function exists only because of that.
+    *)
+    let assoc = Uri.query uri |> List.sort ~cmp:ksrt
+    in
+    List.map
+      ~f:(fun (k,v) -> Printf.sprintf "%s=%s" k (String.concat ~sep:"," v))
+      assoc
+    |> String.concat ~sep:"&"
+
+  let format_time t =
+    (* Core.Std.Time doesn't have a format function that takes a timezone *)
+    let d, s = Time.to_date_ofday ~zone:Time.Zone.utc t in
+    let open Core.Span.Parts in
+    let {hr; min; sec; _} = Time.Ofday.to_parts s in
+    Printf.sprintf "%sT%.2d%.2d%.2dZ"
+      (Date.to_string_iso8601_basic d) hr min sec
+end
+
+type region = [
+  | `ap_northeast_1 (* Asia Pacific (Tokyo) *)
```

Comment:
We don't -- we do need some way to take in the user's input as a string. I was conflicted about including it. Given that this is a short example I'll nix it.